### PR TITLE
DOC: rework documentation for the __astropy_table__ protocol to avoid mingling with numpy 2.0's copy semantics

### LINKED ISSUE #16235: DOC: revise `__astropy_table__`'s documentation ?
I noticed that `__astropy_table__`'s documentation is using a `strict_copy` boolean argument as an example on an additional keyword argument that can be added on the side of the implementor.
See https://docs.astropy.org/en/stable/table/construct_table.html#table-like-objects

In the context of NumPy's change of semantics regarding the `copy` argument (see  #16167 for the broader discussion), it seems to me that this example will not age well and might become confusing (since the upcoming convention, dictated by numpy, is that `copy` arguments are *already* strict). We might also want to update the `__astropy_table__` protocol definition so that the meaning of the `copy` argument aligns with the rest of the code base (i.e., follows numpy at runtime).

In short, I see two documentation changes to be made:
- clarify that the `copy` argument should follow numpy's semantic at runtime (and demonstrate how)
- use a different example for an additional keyword argument

I feel this needs to be addressed in astropy 6.1.0, though I think (?) it could be done past the feature freeze (scheduled for this Friday). Just in case though, let me ping a couple stake holders here: @astrofrog @saimn @mhvk @taldcroft 
To be clear, I can absolutely deal with this before Friday, I'd like to know whether that's required or even desired. Thanks !

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description

Fixes #16235

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: docs/table/construct_table.rst [hunk 1]
```diff
@@ -1244,19 +1244,13 @@ the ``copy`` preference of the original call to ``QTable()``::
...
-If ``copy`` is `True` then the ``__astropy_table__()`` method must ensure that
-a copy of the original data is returned. If ``copy`` is `False` then a
-reference to the table data should be returned if possible. If it is not
-possible (e.g., the original data are in a Python list or must be otherwise
-transformed in memory) then ``__astropy_table__()`` method is free to either
-return a copy or else raise an exception. This choice depends on the preference
-of the implementation. The implementation might choose to allow an additional
-keyword argument (e.g., ``strict_copy`` which gets passed via ``**kwargs``) to
-control the behavior in this case.
+The ``copy`` argument should follow the conventions used by |Table|
```
Comment (taldcroft) [inline docs/table/construct_table.rst:1251]: Missing a trailing period.

File: docs/table/construct_table.rst [hunk 2]
```diff
@@ -1303,15 +1297,7 @@ breakage in this case. ::
...
-          # If a column is a Python list or tuple then it must be copied for
-          # representation in an astropy Table.
-
-          if not copy and strict_copy:
-              for name, col in zip(names, cols):
-                  if not isinstance(col, np.ndarray):
-                      raise ValueError(f'cannot have copy=False because column {name} is '
-                                       'not an ndarray')
+          if mask_invalid:
+              cols = [Masked(col, mask=np.isnan(col)) for col in cols]
```
Comment (taldcroft) [inline docs/table/construct_table.rst:1301]: Even though this is a toy example, it seems like you should only make cols `Masked` if need be:
```suggestion
              cols = [
                  Masked(col, mask=mask) if np.any(mask := np.isnan(col)) else col 
                  for col in cols
              ]
```

**Review by taldcroft — CHANGES_REQUESTED** [review]

File: docs/table/construct_table.rst (hunk 1 above)
Comment (neutrinoceros) [reply docs/table/construct_table.rst]: fixed

File: docs/table/construct_table.rst (hunk 2 above)
Comment (neutrinoceros) [reply docs/table/construct_table.rst]: good point, and love that walrus !

File: docs/table/construct_table.rst [hunk 3]
```diff
@@ -1244,19 +1244,13 @@ the ``copy`` preference of the original call to ``QTable()``::
...
-If ``copy`` is `True` then the ``__astropy_table__()`` method must ensure that
-a copy of the original data is returned. If ``copy`` is `False` then a
-reference to the table data should be returned if possible. If it is not
-possible (e.g., the original data are in a Python list or must be otherwise
-transformed in memory) then ``__astropy_table__()`` method is free to either
-return a copy or else raise an exception. This choice depends on the preference
-of the implementation. The implementation might choose to allow an additional
-keyword argument (e.g., ``strict_copy`` which gets passed via ``**kwargs``) to
-control the behavior in this case.
+The ``copy`` argument should follow the conventions used by |Table|.
```
Comment (mhvk) [inline docs/table/construct_table.rst:1251]: I think we should just delete the paragraph altogether (i.e., also delete 1251-1253 now). We already say `copy` is passed through, and have just mentioned that `mask_invalid` ends up in `**kwargs`.

File: docs/table/construct_table.rst [hunk 4]
```diff
@@ -1303,15 +1297,10 @@ breakage in this case. ::
...
-          # If a column is a Python list or tuple then it must be copied for
-          # representation in an astropy Table.
-
-          if not copy and strict_copy:
-              for name, col in zip(names, cols):
-                  if not isinstance(col, np.ndarray):
-                      raise ValueError(f'cannot have copy=False because column {name} is '
-                                       'not an ndarray')
+          if mask_invalid:
+            cols = [
```
Comment (mhvk) [inline docs/table/construct_table.rst:1301]: Indentation off by 2

**Review by mhvk — COMMENTED** [review]
Thanks, looks good, just a further suggestion for shrinkage, and an indentation nitpick.

File: docs/table/construct_table.rst (hunk 3 above)
Comment (neutrinoceros) [reply docs/table/construct_table.rst]: good call. Fixed

**Review by taldcroft — APPROVED** [review]
Looks good!

**Pull request merged**